nomad/structs/csi_test.go

```diff
@@ -8,41 +8,472 @@ import (
+	claim.Mode = CSIVolumeClaimRead
+	claim.AccessMode = CSIVolumeAccessModeMultiNodeReader
+	claim.AttachmentMode = CSIVolumeAttachmentModeFilesystem
+	require.NoError(vol.Claim(claim, alloc1))
+	require.True(vol.ReadSchedulable())
+	require.False(vol.WriteSchedulable())
+	require.False(vol.WriteFreeClaims())
+	require.Len(vol.ReadClaims, 1)
+	require.Len(vol.WriteClaims, 0)
+	require.Len(vol.PastClaims, 0)
+	require.Equal(CSIVolumeAccessModeMultiNodeReader, vol.AccessMode)
+	require.Equal(CSIVolumeAttachmentModeFilesystem, vol.AttachmentMode)
+	require.Len(vol.RequestedCapabilities, 2)
+	require.Equal(CSIVolumeAccessModeMultiNodeSingleWriter,
+		vol.RequestedCapabilities[0].AccessMode)
+	require.Equal(CSIVolumeAccessModeMultiNodeReader,
+		vol.RequestedCapabilities[1].AccessMode)
+
+	// claim a write and ensure we can't upgrade capabilities.
+	claim.AccessMode = CSIVolumeAccessModeMultiNodeSingleWriter
+	claim.Mode = CSIVolumeClaimWrite
+	claim.AllocationID = alloc2.ID
+	require.EqualError(vol.Claim(claim, alloc2), "unschedulable")
+	require.True(vol.ReadSchedulable())
+	require.False(vol.WriteSchedulable())
+	require.False(vol.WriteFreeClaims())
+	require.Len(vol.ReadClaims, 1)
+	require.Len(vol.WriteClaims, 0)
+	require.Len(vol.PastClaims, 0)
+	require.Equal(CSIVolumeAccessModeMultiNodeReader, vol.AccessMode)
+	require.Equal(CSIVolumeAttachmentModeFilesystem, vol.AttachmentMode)
+
+	// release our last claim, including unpublish workflow
+	claim.AllocationID = alloc1.ID
+	claim.Mode = CSIVolumeClaimRead
+	claim.State = CSIVolumeClaimStateReadyToFree
+	vol.Claim(claim, nil)
+	require.Len(vol.ReadClaims, 0)
+	require.Len(vol.WriteClaims, 0)
+	require.Equal(CSIVolumeAccessModeUnknown, vol.AccessMode)
+	require.Equal(CSIVolumeAttachmentModeUnknown, vol.AttachmentMode)
+	require.Len(vol.RequestedCapabilities, 2)
+	require.Equal(CSIVolumeAccessModeMultiNodeSingleWriter,
+		vol.RequestedCapabilities[0].AccessMode)
+	require.Equal(CSIVolumeAccessModeMultiNodeReader,
+		vol.RequestedCapabilities[1].AccessMode)
+
+	// claim a write on the now-unclaimed volume and ensure we can upgrade
+	// capabilities so long as they're in our RequestedCapabilities.
+	claim.AccessMode = CSIVolumeAccessModeMultiNodeSingleWriter
+	claim.Mode = CSIVolumeClaimWrite
+	claim.State = CSIVolumeClaimStateTaken
+	claim.AllocationID = alloc2.ID
+	require.NoError(vol.Claim(claim, alloc2))
+	require.Len(vol.ReadClaims, 0)
+	require.Len(vol.WriteClaims, 1)
+	require.Equal(CSIVolumeAccessModeMultiNodeSingleWriter, vol.AccessMode)
+	require.Equal(CSIVolumeAttachmentModeFilesystem, vol.AttachmentMode)
+	require.Len(vol.RequestedCapabilities, 2)
+	require.Equal(CSIVolumeAccessModeMultiNodeSingleWriter,
+		vol.RequestedCapabilities[0].AccessMode)
+	require.Equal(CSIVolumeAccessModeMultiNodeReader,
+		vol.RequestedCapabilities[1].AccessMode)
+
+	// claim a write. ensure we no longer have free write claims and that
+	// we've upgraded the volume's access mode (to one that's allowed)
```

Comment:
i don't understand what this is testing, compared to the claim above... is it idempotency? it's not clear to me that this is an upgrade.

Reply:
just for my edification, why is this volume still read-schedulable?

Reply:
> i don't understand what this is testing, compared to the claim above... is it idempotency? it's not clear to me that this is an upgrade.

Good eye... this whole section was a copy-and-paste error that works anyways because it's updating an existing claim. That seems worth keeping, with a corrected docstring.

> just for my edification, why is this volume still read-schedulable?

It's still read-schedulable because `CSIVolumeAccessModeMultiNodeSingleWriter` means "single writer, multiple readers".

Reply:
and (again, begging your patience), why is it still write-schedulable?

Reply:
It's not just you... the API we made here is honestly a little weird and could use some refactoring. `WriteSchedulable` only provides the indication of plugin health and capability, whereas `WriteFreeClaims` is what says "and you can actually claim this volume, given the number of claims it has".